benchmark/utils/timer.hpp

```diff
@@ -0,0 +1,390 @@
+    void toc()
+    {
+        assert(tic_called_ == true);
+        auto ns = this->toc_impl();
+        tic_called_ = false;
+        this->add_record(ns);
+    }
+
+    /**
+     * Get the summation of each time in seconds.
+     *
+     * @return the seconds of total time
+     */
+    double get_total_time() const { return total_duration_sec_; }
+
+    /**
+     * Get the number of repetitions.
+     *
+     * @return the number of repetitions
+     */
+    std::int64_t get_num_repetitions() const { return duration_sec_.size(); }
+
+    /**
+     * Compute the average time of repetitions in seconds
+     *
+     * @return the average time in seconds
+     */
+    double compute_average_time() const
+    {
+        return this->get_total_time() / this->get_num_repetitions();
+    }
+
+    /**
+     * Get the vector containing the time of each repetition in seconds.
+     *
+     * @return the vector of time for each repetition in seconds
+     */
+    std::vector<double> get_time_detail() const { return duration_sec_; }
+
+    /**
+     * Get the latest result in seconds. If there is no result yet, return
+     * 0.
+     *
+     * @return the latest result in seconds
+     */
+    double get_latest_time() const
+    {
+        if (duration_sec_.size() >= 1) {
+            return duration_sec_.back();
+        } else {
+            return 0;
+        }
+    }
+
+    /**
+     * Clear the results of timer
+     */
+    void clear()
+    {
+        duration_sec_.clear();
+        tic_called_ = false;
+        total_duration_sec_ = 0;
+    }
+
+    /**
+     * Create a timer
+     */
+    Timer() : tic_called_(false), total_duration_sec_(0) {}
+
+protected:
+    /**
+     * Put the second result into vector
+     *
+     * @param ns  the second result to insert
+     */
+    void add_record(double ns)
+    {
+        // add the result;
+        duration_sec_.emplace_back(ns);
+        total_duration_sec_ += ns;
+    }
+
+    /**
+     * The implementation of tic.
+     */
+    virtual void tic_impl() = 0;
+
+    /**
+     * The implementation of toc. Return the seconds result.
+     *
+     * @return the seconds result
+     */
+    virtual double toc_impl() = 0;
+
+private:
+    std::vector<double> duration_sec_;
+    bool tic_called_;
+    double total_duration_sec_;
+};
+
+
+/**
+ * CpuTimer uses the synchronize of the executor and std::chrono to measure the
+ * timing.
+ */
+class CpuTimer : public Timer {
+public:
+    /**
+     * Create a CpuTimer
+     *
+     * @param exec  Executor associated to the timer
+     */
+    CpuTimer(std::shared_ptr<const gko::Executor> exec) : Timer(), exec_(exec)
+    {}
+
+protected:
+    void tic_impl() override
+    {
+        exec_->synchronize();
+        start_ = std::chrono::steady_clock::now();
+    }
+
+    double toc_impl() override
+    {
+        exec_->synchronize();
+        auto stop = std::chrono::steady_clock::now();
+        auto duration_time = get_duration_in_seconds(stop - start_);
```

Comment:
We could just use the standard way of doing this as shown here: https://en.cppreference.com/w/cpp/chrono/steady_clock/now
We don't really need an entire function and file to do this.
```suggestion
        std::chrono::duration<double> duration_time = stop - start_;
```

Reply:
thanks for the example, I did not know that before.